server/src/main/java/org/elasticsearch/common/logging/DeprecationIndexingService.java

```diff
@@ -0,0 +1,154 @@
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.elasticsearch.common.Strings.isNullOrEmpty;
+
+/**
+ * This service is responsible for writing deprecation messages to a data stream. It also creates
+ * the data stream if necessary. The writing of messages can be toggled using the
+ * {@link #WRITE_DEPRECATION_LOGS_TO_INDEX} setting.
+ */
+public class DeprecationIndexingService implements ClusterStateListener {
+    private static final Logger LOGGER = LogManager.getLogger(DeprecationIndexingService.class);
+
+    private static final String DATA_STREAM_NAME = "logs-deprecation-elasticsearch";
+    private static final String TEMPLATE_NAME = DATA_STREAM_NAME + "-template";
+    private static final String TEMPLATE_MAPPING = TEMPLATE_NAME + ".json";
+
+    private static final String DEPRECATION_ORIGIN = "deprecation";
+
+    public static final Setting<Boolean> WRITE_DEPRECATION_LOGS_TO_INDEX = Setting.boolSetting(
+        "cluster.deprecation_indexing.enabled",
+        true,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    private final Client client;
+    private boolean hasTriedToLoadTemplate = false;
+    private volatile boolean isEnabled = false;
+
+    public DeprecationIndexingService(ClusterService clusterService, Client client) {
+        this.client = new OriginSettingClient(client, DEPRECATION_ORIGIN);
+
+        clusterService.addListener(this);
+    }
+
+    /**
+     * Indexes a deprecation message.
+     * @param key       the key that was used to determine if this deprecation should have been be logged.
+     *                  Useful when aggregating the recorded messages.
+     * @param message   the message to log
+     * @param xOpaqueId the associated "X-Opaque-ID" header value if any, or <code>null</code>
+     * @param params    parameters to the message, if any
+     */
+    public void writeMessage(String key, String message, String xOpaqueId, Object[] params) {
+        if (this.isEnabled == false) {
+            return;
+        }
+
+        Map<String, Object> payload = new HashMap<>();
+        payload.put("@timestamp", Instant.now().toString());
+        payload.put("key", key);
+        payload.put("message", message);
+
+        if (isNullOrEmpty(xOpaqueId) == false) {
+            payload.put("x-opaque-id", xOpaqueId);
+        }
+
+        if (params != null && params.length > 0) {
+            payload.put("params", params);
+        }
+
+        new IndexRequestBuilder(client, IndexAction.INSTANCE).setIndex(DATA_STREAM_NAME)
```

Comment:
I would suggest to use the BulkProcessor (https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java) since it allows for a configurable flush interval and optional incremental backoff retry logic.